revoke.go

```diff
@@ -0,0 +1,81 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"github.com/coreos/go-oidc"
+	"github.com/pkg/errors"
+	"golang.org/x/oauth2"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strings"
+)
+
+// revocationEndpoint parses the OIDC Provider claims from the discovery document
+// and tries to find the revocation_endpoint.
+func revocationEndpoint(p *oidc.Provider) (string, error) {
+	claims := struct {
+		RevocationEndpoint string `json:"revocation_endpoint"`
+	}{}
+	if err := p.Claims(&claims); err != nil {
+		return "", errors.Wrap(err, "Error unmarshalling provider doc into struct")
+	}
+	if claims.RevocationEndpoint == "" {
+		return "", errors.New("Provider doesn't have a revocation_endpoint")
+	}
+	return claims.RevocationEndpoint, nil
+}
+
+// revokeTokens is a helper that takes an oauth2.Token and revokes the access and refresh tokens.
+// If no tokens are found, it succeeds.
+func revokeTokens(ctx context.Context, revocationEndpoint string, token *oauth2.Token, clientID, clientSecret string) error {
+	if token.AccessToken != "" {
+		err := revokeToken(ctx, revocationEndpoint, token.AccessToken, "access_token", clientID, clientSecret)
+		if err != nil {
+			return errors.Wrap(err, "Failed to revoke access token")
+		}
+	}
+	if token.RefreshToken != "" {
+		err := revokeToken(ctx, revocationEndpoint, token.AccessToken, "refresh_token", clientID, clientSecret)
+		if err != nil {
+			return errors.Wrap(err, "Failed to revoke refresh token")
+		}
+	}
+	return nil
+}
+
+// revokeToken takes care of revoking an access/refresh token to the IdP.
+// The revocation procedure is described in RFC7009:+
+// https://tools.ietf.org/html/rfc7009
+func revokeToken(ctx context.Context, revocationEndpoint string, token, tokenType, clientID, clientSecret string) error {
+	// Verify revocation_endpoint has https url
+	if !strings.HasPrefix(revocationEndpoint, "https") {
+		return errors.New(fmt.Sprintf("Revocation endpoint (%v) MUST use https", revocationEndpoint))
+	}
+	values := url.Values{}
+	values.Set("token", token)
+	values.Set("token_type_hint", tokenType)
+	req, err := http.NewRequest(http.MethodPost, revocationEndpoint, strings.NewReader(values.Encode()))
+	if err != nil {
+		return err
+	}
+	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+	// Assume provider accepts http basic auth, because it is REQUIRED by the spec
+	req.SetBasicAuth(clientID, clientSecret)
```

Comment:
I think this does not always hold. From https://tools.ietf.org/html/rfc6749#section-2.3

>    If the client type is confidential, the client and authorization
>    server establish a client authentication method suitable for the
>    security requirements of the authorization server.  The authorization
>    server MAY accept any form of client authentication meeting its
>    security requirements.

Surely, the most common authentication method is via the `Authentication` header, but there are other schemes as well, such as the `X-Auth-Token` of Keystone. I'm not saying we need to change this now, but we should have this in the back of our head. E.g., we could extend the provider claims to specify the authentication header they use.

Reply:
Hmm I think you are right. I was looking at the client password section that says:
> The authorization server MUST support the HTTP Basic
   authentication scheme for authenticating clients that were issued a
   client password.

But of course, there are other methods of authentication.
It's interesting to see what Go's oauth2 package does for this scenario. In auth_code <-> token exchange, where client authentication is needed, they try both basic auth AND posting credentials and cache whatever ends up working:
https://github.com/golang/oauth2/blob/bf48bf16ab8d622ce64ec6ce98d2c98f916b6303/internal/token.go#L204-L215

I will add a note that we may need to support more in the future.